[export] Fix unflattened submodule ordering.

zhxchen17 opened the pull request:
Summary:
Make sure the order of submodules is the same as the original eager module.

bypass-github-export-checks

Test Plan: buck test mode/opt caffe2/test:test_export -- -r test_unflatten_submodule_ordering

Differential Revision: D55151837

@diff-train-skip-merge

avikchaudhuri commented on test/export/test_unflatten.py:
```diff
@@ -615,6 +615,46 @@ def forward(self, x):
+        unflattened = torch.export.unflatten(ep)
+        fqn_list = [x for x, _ in unflattened.named_modules(remove_duplicate=False)]
+        self.assertEqual(len(fqn_list), 4)
+        self.assertEqual(
```
nice

avikchaudhuri commented on torch/export/unflatten.py:
```diff
@@ -788,6 +791,20 @@ def _outline_submodules(orig_graph: torch.fx.Graph, root_module: UnflattenedModu
+        if child is None:
+            continue
+        fqn = prefix + name
+        _reorder_submodules(child, fqn_order, fqn + ".")
```
add `prefix=` to this call?

avikchaudhuri commented on torch/export/unflatten.py:
```diff
@@ -788,6 +791,20 @@ def _outline_submodules(orig_graph: torch.fx.Graph, root_module: UnflattenedModu
+            continue
+        fqn = prefix + name
+        _reorder_submodules(child, fqn_order, fqn + ".")
+        delattr(parent, name)
```
Is this needed because `register_module` will error otherwise?

zhxchen17 replied:
yes

zhxchen17 committed:
[export] Fix unflattened submodule ordering. (#122341)

Summary:

Make sure the order of submodules is the same as the original eager module.

bypass-github-export-checks

Test Plan: buck test mode/opt caffe2/test:test_export -- -r test_unflatten_submodule_ordering

Reviewed By: avikchaudhuri

Differential Revision: D55151837

huydhn commented:
@pytorchbot merge

malfet commented:
Cancelled the merge as I had to revert it internally

huydhn commented:
Although the diff has been reverted, ShipIt is still failing on this PR.  So, I add @diff-train-skip-merge and close this one. It's better to create a new PR/diff to reland this change.
